### PR TITLE
Use a callback to close TCP Comms, rather than check every time

### PR DESCRIPTION
**Pull request by mrocklin**
In a recent trace this relieves about 30ms of a 3s shuffle computation
resulting in around a 1% overall speedup

cc @jakirkham this might help a bit with the worker_send/client_send code in transition.  I think that the real solution is still to try to batch things there though.

### CODE REVIEW

**Review by jakirkham — APPROVED** [review]

**Comment by mrocklin** [comment]
I recommend that we merge this after the upcoming release.  cc @jrbourbeau

**Comment by jakirkham** [comment]
Sorry Matt maybe I should have merged this earlier. Am ok with waiting until after the release as well

**Pull request merged**
